internal/loader/configuration_client_manager.go

```diff
@@ -0,0 +1,424 @@
+	EndpointSection                     string        = "Endpoint"
+	SecretSection                       string        = "Secret"
+	IdSection                           string        = "Id"
+	AzConfigDomainLabel                 string        = "azconfig"
+	AppConfigDomainLabel                string        = "appconfig"
+	FallbackClientRefreshExpireInterval time.Duration = time.Hour
+	MinimalClientRefreshInterval        time.Duration = time.Second * 30
+	MaxBackoffDuration                  time.Duration = time.Minute * 10
+	MinBackoffDuration                  time.Duration = time.Second * 30
+	JitterRatio                         float64       = 0.25
+	SafeShiftLimit                      int           = 63
+)
+
+var (
+	clientOptionWithModuleInfo *azappconfig.ClientOptions = &azappconfig.ClientOptions{
+		ClientOptions: policy.ClientOptions{
+			Telemetry: policy.TelemetryOptions{
+				ApplicationID: fmt.Sprintf("%s/%s", properties.ModuleName, properties.ModuleVersion),
+			},
+		},
+	}
+)
+
+func NewConfigurationClientManager(ctx context.Context, provider acpv1.AzureAppConfigurationProvider) (*ConfigurationClientManager, error) {
+	manager := &ConfigurationClientManager{
+		ReplicaDiscoveryEnabled: provider.Spec.ReplicaDiscovery,
+		ValidDomain:             getValidDomain(*provider.Spec.Endpoint),
+		Endpoint:                *provider.Spec.Endpoint,
+		ClientsFallbackStatus:   make(map[string]*ConfigurationClientWrapper),
+	}
+
+	var err error
+	if provider.Spec.ConnectionStringReference != nil {
+		manager.ConnectionString, err = getConnectionStringParameter(ctx, types.NamespacedName{Namespace: provider.Namespace, Name: *provider.Spec.ConnectionStringReference})
+		if err != nil {
+			return nil, err
+		}
+		manager.secret, err = parseConnectionString(manager.ConnectionString, SecretSection)
+		if err != nil {
+			return nil, err
+		}
+		manager.id, err = parseConnectionString(manager.ConnectionString, IdSection)
+		if err != nil {
+			return nil, err
+		}
+
+		manager.SpecifiedClient, err = azappconfig.NewClientFromConnectionString(manager.ConnectionString, clientOptionWithModuleInfo)
+		if err != nil {
+			return nil, err
+		}
+
+		return manager, nil
+	}
+
+	manager.Credential, err = CreateTokenCredential(ctx, provider.Spec.Auth, provider.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	manager.SpecifiedClient, err = azappconfig.NewClient(*provider.Spec.Endpoint, manager.Credential, clientOptionWithModuleInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return manager, nil
+}
+
+func SrvTargetHostQuery(host string) ([]string, error) {
+	results := make([]string, 0)
+
+	_, originRecords, err := net.LookupSRV(Origin, TCP, host)
+	if err != nil {
+		// If the host does not have SRV records => no replicas
+		if _, ok := err.(*net.DNSError); ok {
+			results = append(results, host)
+			return results, nil
+		}
+		return nil, err
+	}
+
+	originHost := strings.TrimSuffix(originRecords[0].Target, ".")
```

Comment:
just want to make sure,  the originRecords will never be an empty array. right?

Reply:
Each replica has `_origin._tcp.*` SRV record been created to indicate the origin store of the replica. If there's no replica of user's appconfig store, the originRecoreds will be empty.